components/director/pkg/operation/k8s/k8s_scheduler.go

```diff
@@ -0,0 +1,94 @@
+package k8s
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/kyma-incubator/compass/components/director/pkg/operation"
+	"github.com/kyma-incubator/compass/components/operations-controller/api/v1alpha1"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+//go:generate mockery -name=K8SClient -output=automock -outpkg=automock -case=underscore
+type K8SClient interface {
+	Create(ctx context.Context, operation *v1alpha1.Operation) (*v1alpha1.Operation, error)
+	Get(ctx context.Context, name string, options metav1.GetOptions) (*v1alpha1.Operation, error)
+	Update(ctx context.Context, operation *v1alpha1.Operation) (*v1alpha1.Operation, error)
+}
+
+type Scheduler struct {
+	kcli K8SClient
+}
+
+func NewScheduler(kcli K8SClient) *Scheduler {
+	return &Scheduler{
+		kcli: kcli,
+	}
+}
+
+func (s *Scheduler) Schedule(ctx context.Context, op *operation.Operation) (string, error) {
+	operationName := fmt.Sprintf("%s-%s", op.ResourceType, op.ResourceID)
+	getOp, err := s.kcli.Get(ctx, operationName, metav1.GetOptions{})
+	if err != nil {
+		if errors.IsNotFound(err) {
+			k8sOp := toK8SOperation(op)
+			if createdOperation, err := s.kcli.Create(ctx, k8sOp); err != nil {
+				return "", err
+			} else {
+				return string(createdOperation.UID), nil
+			}
+		}
+		return "", err
+	}
+	if isOpInProgress(getOp) {
+		return "", fmt.Errorf("another operation is in progress for resource with ID %q", op.ResourceID)
+	}
+	getOp = updateOperationSpec(op, getOp)
+	updatedOperation, err := s.kcli.Update(ctx, getOp)
+	if err != nil {
+		if errors.IsConflict(err) {
+			return "", fmt.Errorf("another operation is in progress for resource with ID %q", op.ResourceID)
+		}
+		return "", err
+	}
+	return string(updatedOperation.UID), err
+}
+
+func isOpInProgress(op *v1alpha1.Operation) bool {
+	for _, cond := range op.Status.Conditions {
+		if cond.Status == v1.ConditionTrue {
+			return false
+		}
+	}
+	return true
+}
+
+func toK8SOperation(op *operation.Operation) *v1alpha1.Operation {
+	operationName := fmt.Sprintf("%s-%s", op.ResourceType, op.ResourceID)
+	result := &v1alpha1.Operation{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: operationName,
+		},
+	}
+	return updateOperationSpec(op, result)
+}
+
+func updateOperationSpec(op *operation.Operation, k8sOp *v1alpha1.Operation) *v1alpha1.Operation {
+	k8sOp.Spec = v1alpha1.OperationSpec{
+		OperationCategory: op.OperationCategory,
+		OperationType:     v1alpha1.OperationType(title(string(op.OperationType))),
```

Comment:
This part here - `title(string(op.OperationType))`, will cause the new Operations update handler to fail validating the request as it expects all upper case letters for the operation type:

```
if err := validation.ValidateStruct(&operation,
	validation.Field(&operation.ResourceID, is.UUID),
	validation.Field(&operation.OperationType, validation.Required, validation.In(graphql.OperationTypeCreate, graphql.OperationTypeUpdate, graphql.OperationTypeDelete)),
	validation.Field(&operation.ResourceType, validation.Required, validation.In(resource.Application))); err != nil {
	apperrors.WriteAppError(ctx, writer, apperrors.NewInvalidDataError("Invalid operation properties: %s", err), http.StatusBadRequest)
	return
}
```

Reference link: https://github.com/kyma-incubator/compass/blob/5debfe3ffcc4e31a3487995b17f25e53435a9eff/components/director/pkg/operation/update_operation_handler.go#L98

Reply:
We'll need to tweak the validation then, good catch! The "title"-thingy was needed because of the OperationType in the CRD: https://github.com/kyma-incubator/compass/blob/22003412f0640de76ce17b07fcbc424964e6e6c6/components/operations-controller/api/v1alpha1/operation_types.go#L22

Reply:
It started to look really workaround-ish and I've just changed the types in the `graphql` package to match the ones in the CRD